semantic-tags/src/Tags/Tagging/Precise.hs

```diff
@@ -43,36 +64,133 @@ firstLine :: Source -> Text
+
+
+-- | This function may be used as a value for 'foldMap' in a 'Foldable' instance.
+foldMapDefault1 :: (Traversable1 Foldable t, Monoid b) => (a -> b) -> t a -> b
+foldMapDefault1 f = foldMap1 @Foldable f (foldMap f)
+
+-- | This function may be used as a value for 'fmap' in a 'Functor' instance.
+fmapDefault1 :: Traversable1 Functor t => (a -> b) -> t a -> t b
+fmapDefault1 f = runIdentity . traverse1 @Functor (Identity . f) (Identity . fmap f)
+
+-- | This function may be used as a value for 'traverse' in a 'Traversable' instance.
+traverseDefault1 :: (Traversable1 Traversable t, Applicative f) => (a -> f b) -> t a -> f (t b)
+traverseDefault1 f = traverse1 @Traversable f (traverse f)
+
+
+-- FIXME: move GTraversable1 into semantic-ast.
+class GTraversable1 c t where
+  -- | Generically map annotations and subterms of kind @* -> *@ into an 'Applicative' context.
+  gtraverse1
+    :: Applicative f
+    => (a -> f b)
+    -> (forall t' . c t' => t' a -> f (t' b))
+    -> t a
+    -> f (t b)
+
+instance GTraversable1 c f => GTraversable1 c (M1 i c' f) where
+  gtraverse1 f g = fmap M1 . gtraverse1 @c f g . unM1
+
+instance (GTraversable1 c f, GTraversable1 c g) => GTraversable1 c (f :*: g) where
+  gtraverse1 f g (l :*: r) = (:*:) <$> gtraverse1 @c f g l <*> gtraverse1 @c f g r
+
+instance (GTraversable1 c f, GTraversable1 c g) => GTraversable1 c (f :+: g) where
+  gtraverse1 f g (L1 l) = L1 <$> gtraverse1 @c f g l
+  gtraverse1 f g (R1 r) = R1 <$> gtraverse1 @c f g r
+
+instance GTraversable1 c (K1 R t) where
+  gtraverse1 _ _ (K1 k) = pure (K1 k)
+
+instance GTraversable1 c Par1 where
+  gtraverse1 f _ (Par1 a) = Par1 <$> f a
+
+instance c t => GTraversable1 c (Rec1 t) where
+  gtraverse1 _ g (Rec1 t) = Rec1 <$> g t
 
-instance GFoldable1 c f => GFoldable1 c (M1 i c' f) where
-  gfoldMap1 alg = gfoldMap1 @c alg . unM1
+instance (Traversable f, GTraversable1 c g) => GTraversable1 c (f :.: g) where
+  gtraverse1 f g = fmap Comp1 . traverse (gtraverse1 @c f g) . unComp1
 
-instance (GFoldable1 c f, GFoldable1 c g) => GFoldable1 c (f :*: g) where
-  gfoldMap1 alg (f :*: g) = gfoldMap1 @c alg f <> gfoldMap1 @c alg g
+instance GTraversable1 c U1 where
+  gtraverse1 _ _ _ = pure U1
 
-instance (GFoldable1 c f, GFoldable1 c g) => GFoldable1 c (f :+: g) where
-  gfoldMap1 alg (L1 l) = gfoldMap1 @c alg l
-  gfoldMap1 alg (R1 r) = gfoldMap1 @c alg r
 
-instance GFoldable1 c (K1 R t) where
-  gfoldMap1 _ _ = mempty
+-- | @'Generics' t@ has a 'Traversable1' instance when @'Rep1' t@ has a 'GTraversable1' instance, making this convenient for applying 'traverse1' to 'Generic1' types lacking 'Traversable1' instances:
+--
+-- @
+-- 'getGenerics' '<$>' 'traverse1' f g ('Generics' t) = 'to1' '<$>' 'gtraverse1' f g ('from1' t)
+-- @
+--
+-- It further defines its 'Foldable', 'Functor', and 'Traversable' instances using 'Traversable1', making it suitable for deriving with @-XDerivingVia@.
+newtype Generics t a = Generics { getGenerics :: t a }
```

Comment:
This, on the other hand, is kind of fun.

We don’t have `Traversable1` instances for our AST types yet, so this bridges the gap—wrap a `Generic1` term in `Generics` and you can use `Traversable1` by means of said instance.

It also supports use of `-XDerivingVia` (for classes supporting `-XGeneralizedNewtypeDeriving`… alas, this does not include `Traversable` or `Traversable1`), which is what prompted the name. I intend to measure how long it takes to derive `Foldable` & `Functor` using `-XDeriveFoldable`/`-XDeriveFunctor`, vs. how long it takes using `-XDerivingVia` for our TH-generated AST datatypes:

```haskell
newtype Expr a = Expr ((Thing :+: OtherThing) a)
  deriving (Foldable, Functor) via Generics
```